### ISSUE TITLE
A proper way to create wrapper for group and view

### ISSUE THREAD

**Issue by Diarica** [skypjack/entt#1176]
`class VTransformSystem : public VSystem
{
public:

    void OnStartup(VEntityManager& manager) override;
    void OnExecute(VEntityManager& manager, const VSystemExecutionContext& context) override;
    
    void SetPosition(VTransformComponent& tran, const glm::vec3& position);
    void SetRotation(VTransformComponent& tran, const glm::quat& rotation);
    void SetScale(VTransformComponent& tran, const glm::vec3& scale);
    
    static glm::mat4 CalculateTransformMatrix(const glm::vec3& position, const glm::quat& rotation, const glm::vec3& scale);

    
    auto trangroup = // 
};
`

I'm building a ECS inside my game engine using entt, for some particular reason, I can't just create groups just on the right hand side of group declaration like auto group = reg.group.... (there is multiple registry supported in my game engine, just like Worlds in Unity DOTS)
std::optional is not working for this situation.

Any helps will be thanksful.

**Comment by Diarica** [comment]
> I'd like to help you but I've to admit that I don't understand what your problem is. Can you provide more details? Does the `registry.group` call not work for you or do you just want to create an _empty_ group to initialize later?

I'm sorry for unclear expression.
Yes, I just want to create an empty group and initialize later, and before you replied me, I found as_view could work.
But I originally want to create a wrapper of view and groups, Is there a fesiable approarch?
Wrappers could just be type alias of entt things, like this
`template<typename Get, typename Exclude>
using View = entt::view<Get, Exclude>;

template<typename Owned, typename Get, typename Exclude>
using Group = entt::group<Owned, Get, Exclude>;`

Really thanks for your helps.

**Comment by skypjack** [comment]
It's more similar to the following then:
```cpp
template<typename... Owned, typename... Get, typename... Exclude>
using Group = entt::group<entt::type_list<Owned...>, entt::type_list<Get...>, entt::type_list<Exclude...>>;
```
Same for the view type.

**Comment by Diarica** [comment]
> It's more similar to the following then:
> 
> ```c++
> template<typename... Owned, typename... Get, typename... Exclude>
> using Group = entt::group<entt::type_list<Owned...>, entt::type_list<Get...>, entt::type_list<Exclude...>>;
> ```
> 
> Same for the view type.

Thanks, I have already done the wrapper, but I have one last question, in view and group function, auto is the only choice to store the return value? 
`auto test = m_enttRegistry.view<component>();`
Cannot store things like 
`basic_group<owned_t<storage_for_type<const Owned>...>, get_t<storage_for_type<const Get>...>, exclude_t<storage_for_type<const Exclude>...>>` ?

**Comment by skypjack** [comment]
Of course, yes, you can use the explicit type in all cases if you want. That's about C++ though. It doesn't depend on EnTT.

**Comment by Diarica** [comment]
> ```c++
> template<typename... Owned, typename... Get, typename... Exclude>
> ```

I'm sorry, but I have to back the original question, how to create an empty group and initialize it later?

`template<typename... Owned, typename... Get, typename... Exclude>
class VQuery
{

public:
    void InitializeQuery(VEntityManager& entity_manager)
    {
        m_Group = entt::as_group(entity_manager.GetNativeRegistry());
    }
    operator auto&()
    {
        return m_Group;
    }
private:
    entt::group<Owned...,Get...,Exclude...> m_Group;
    
};`

It may like that, but it seems like I cannot access any group functions(like contains(), size(), strorage()... )

**Comment by skypjack** [comment]
> it seems like I cannot access any group functions(like contains(), size(), strorage()... )

What does it mean? Can you provide a minimal example that I can test and fix eventually? I don't even get what you're trying to do nor what the problem is.

**Comment by Diarica** [comment]
> ```c++
> entt::type_list<Owned...>, entt::type_list<Get...>, entt::type_list<Exclude...>
> ```

Okay, sorry, I would like to create a wrapper of entt::group, and it's need be able to initialize group with registry later.


```template<typename... Owned, typename... Get, typename... Exclude>
class VQuery
{

public:
    void InitializeQuery(VEntityManager& entity_manager)
    {
        m_Group = entt::as_group(entity_manager.GetNativeRegistry());
    }
    operator auto&()
    {
        return m_Group;
    }

    auto& GetTest()
    {
        return m_Group;
    }
    
    
private:
    entt::group<entt::type_list<Owned...>, entt::type_list<Get...>, entt::type_list<Exclude...>> m_Group;
    
};
```
I tried this, but it's not work for me, cannot be compiled.

**Comment by skypjack** [comment]
If you want to use `entt::group`, then it's something like this:
```cpp
entt::group<entt::owned_t<Owned...>, entt::get_t<Get...>, entt::exclude_t<Exclude...>> m_Group;
```
As for:

> cannot be compiled

If it doesn't work, please, post the relevant part of your error or a reproducible example. It saves time when helping. 👍

**Comment by Diarica** [comment]
Thanks! You're so kind! Sorry for I get things complicated.
But it's still cannot be compiled.

Here I wrote a simple wrapper class of entt::group, strictly followed your example.
```
template<typename... Owned, typename... Get, typename... Exclude>
class test_query
{
    entt::group<entt::owned_t<Owned...>, entt::get_t<Get...>, entt::exclude_t<Exclude...>> m_Group;
};
```

It's just a simple class with those template parameters and entt::group
but it's still get compile errors, 
`Error C3525 : 'Owned': if a class template has a template parameter pack it must appear at the end of the template parameter list`

I even not sure if it's a C++ template restriction? Class templates only supported single variety parameters?
But type alias actually supported multiple variety template parameters.
Here's an example that I tested for 'using'

`template<typename... Owned, typename... Get, typename... Exclude>
using f = entt::group<entt::owned_t<Owned...>, entt::get_t<Get...>, entt::exclude_t<Exclude...>>;`
It's compiled successfully.

Sorry I'm not familiar with templates, It's really a restriction? (Class templates only supported single variety parameters?)

Is there a feasible way to create a wrapper of entt::group?

**Comment by skypjack** [comment]
Consider this:
```cpp
template<typename... Owned, typename... Get, typename... Exclude>
class test_query;
```
Then imagine the following:
```cpp
test_query<int, char> q;
```
What pack `int` belongs to? `Owned`? `Get`? `Exclude`? What's the one for `char` then?
The compiler cannot know what you want in case of ambiguity. So, yeah, the error is in your code and you should fix it.

**Comment by Diarica** [comment]
> Consider this:
> 
> ```c++
> template<typename... Owned, typename... Get, typename... Exclude>
> class test_query;
> ```
> 
> Then imagine the following:
> 
> ```c++
> test_query<int, char> q;
> ```
> 
> What pack `int` belongs to? `Owned`? `Get`? `Exclude`? What's the one for `char` then? The compiler cannot know what you want in case of ambiguity. So, yeah, the error is in your code and you should fix it.

Understood, but I literally can't just use type alias to wrap entt::group. I need extra behaviors in my wrapper class, because I need to initialize group with registry later, and store it to a std::vector container.

So, Is there a possible way to create a wrapper class with same parameters than call entt::registry::group ?
I mean, if I need to create a wrapped group, I can simply says

`WrappedGroup<Transform,Tag> group;
group.init(getRegistry()); // here to initialize group later. 
group.getNativeGroup().view([]) // here to access entt::group functions`

Thanks for your kindness help again.

**Comment by Innokentiy-Alaytsev** [comment]
Some options:
1. Use opaque type lists in your class declaration, and pass them directly to the wrapped group.
2. Use a factory function for creating your wrapper, similar to how `entt:group`s are created (in most cases?).

**Comment by Diarica** [comment]
> Some options:
> 
> 1. Use opaque type lists in your class declaration, and pass them directly to the wrapped group.
> 2. Use a factory function for creating your wrapper, similar to how `entt:group`s are created (in most cases?).

Did you mean this? I tried to pass same template parameter than entt::basic_group, but it's cannot be compiled. I don't know how do I use that.
```template<typename Owned, typename Get, typename Exclude>
class VQuery : public VQueryBase
{

public:
    void* operator new(std::size_t) = delete;
    void operator delete(void*) = delete;

    
    VQuery(VSystem* owningSys): VQueryBase(owningSys)
    {

    }

    void InitializeQuery(VEntityManager& entity_manager) override
    {
        
        m_Group = entity_manager.group<>();
    }
    
    auto& GetGroup()
    {
        
        return m_Group;
    }


    
private:
    entt::basic_group<entt::type_list_transform_t<Owned, entt::storage_for>, entt::type_list_transform_t<Get, entt::storage_for>, entt::type_list_transform_t<Exclude, entt::storage_for>> m_Group;
    
};```

**Comment by Innokentiy-Alaytsev** [comment]
There is also an option of wrapping an externally created group, i.e. the wrapper can be parametrised by the group type and not component type lists:

```c++
template< class TGroup >
class Wrapper {
 public:
    explicit Wrapper (const TGroup& i_group)
        : the_group {i_group} 
    {
    }


 private:
    TGroup the_group_;
};
```

This version is simpler, but not as encapsulated. Whether you want this level of encapsulation or not depends on you. I believe, in most cases you don't need it.

In any case, consider the price of implementing the version you choose and the maintainability of the implementation.

**Comment by Diarica** [comment]
> The code bellow was not tested and mostly represents the general idea. Some parts, like passing the registry, etc., are omitted for brevity.
> 
> 1. Something like this (I guess it should work to some extent):
> 
> ```c++
> template< class TOwned, class TGet = entt::get_t<>, class TExclude = entt::exclude_t<> >
> class Wrapper {
>  public:
>     // ...
> 
>  private:
>     entt::basic_group<TOwned, TGet, TExclude> the_group_;
> };
> 
> Wrapper< entt::owned_t< CO1, CO2 >, entt::get_t< CG1, CG2, CG3 >, entt::exclude_t< CE1 > > wrapper_1;
> Wrapper< entt::owned_t< CO3, CO4 >, entt::get_t< >, entt::exclude_t< CE1 > > wrapper_2;
> ```
> 
> 2. Duplicate to some extent what EnTT supposedly does somewhere:
> 
> ```c++
> template< class, class, class >
> class Wrapper;
> 
> template<class...TOwneds, class... TGets, class... TExcludes>
> class Wrapper<owned_t<TOwneds...>, get_t<TGets...>, exclude_t<TExcludes...>> {
>  public:
>     // ...
>  private:
>     entt::basic_group< owned_t<TOwneds...>, get_t<TGets...>, exclude_t<TExcludes...>> the_group_;
> };
> 
> template<class...TOwneds, class... TGets, class... TExcludes>
> auto MakeWrapper (owned_t<TOwneds...>, get_t<TGets...>, exclude_t<TExcludes...>) 
> {
>     return Wrapper<owned_t<TOwneds...>, get_t<TGets...>, exclude_t<TExcludes...>>{};
> }
> ```

Thank you! It's a very good apporach, but I still get compile errors, maybe it's because type traits inside entt::registry::group ?

Use case:
`VQuery<entt::owned_t<VTransformComponent>> m_TransformQuery = VQuery<entt::owned_t<VTransformComponent>>(this);`

VQuery Definition
`template< class TOwned, class TGet = entt::get_t<>, class TExclude = entt::exclude_t<>>
class VQuery : public VQueryBase
{
public:
    // ...

    VQuery(VSystem* owningSys): VQueryBase(owningSys)
    {

        
    }

    
    void* operator new(std::size_t) = delete;
    void operator delete(void*) = delete;


    auto& GetGroup()
    {
        entt::registry f;
        
        return m_Group;
    }

private:

    entt::basic_group<TOwned, TGet, TExclude> m_Group;
};

`
 Compile errors
`7>entt.hpp(18783,58): Error C2027 : use of undefined type 'VTransformComponent'
7>entt.hpp(18783,58): Error C2146 : syntax error: missing '>' before identifier 'base_type'`

Type traits inside entt.hpp
`template<typename... Owned, typename... Get, typename... Exclude>
class basic_group<owned_t<Owned...>, get_t<Get...>, exclude_t<Exclude...>> {
    using base_type = std::common_type_t<typename Owned::base_type..., typename Get::base_type..., typename Exclude::base_type...>;`

I have no clue about it, But it's on the right track I believe.

**Comment by Diarica** [comment]
> The code bellow was not tested and mostly represents the general idea. Some parts, like passing the registry, etc., are omitted for brevity.
> 
> 1. Something like this (I guess it should work to some extent):
> 
> ```c++
> template< class TOwned, class TGet = entt::get_t<>, class TExclude = entt::exclude_t<> >
> class Wrapper {
>  public:
>     // ...
> 
>  private:
>     entt::basic_group<TOwned, TGet, TExclude> the_group_;
> };
> 
> Wrapper< entt::owned_t< CO1, CO2 >, entt::get_t< CG1, CG2, CG3 >, entt::exclude_t< CE1 > > wrapper_1;
> Wrapper< entt::owned_t< CO3, CO4 >, entt::get_t< >, entt::exclude_t< CE1 > > wrapper_2;
> ```
> 
> 2. Duplicate to some extent what EnTT supposedly does somewhere:
> 
> ```c++
> template< class, class, class >
> class Wrapper;
> 
> template<class...TOwneds, class... TGets, class... TExcludes>
> class Wrapper<owned_t<TOwneds...>, get_t<TGets...>, exclude_t<TExcludes...>> {
>  public:
>     // ...
>  private:
>     entt::basic_group< owned_t<TOwneds...>, get_t<TGets...>, exclude_t<TExcludes...>> the_group_;
> };
> 
> template<class...TOwneds, class... TGets, class... TExcludes>
> auto MakeWrapper (owned_t<TOwneds...>, get_t<TGets...>, exclude_t<TExcludes...>) 
> {
>     return Wrapper<owned_t<TOwneds...>, get_t<TGets...>, exclude_t<TExcludes...>>{};
> }
> ```


Thank you! It's a very good apporach, but I still get compile errors, maybe it's because type traits inside entt::registry::group ?

Use case : ```VQuery<entt::owned_t<VTransformComponent>> m_TransformQuery = VQuery<entt::owned_t<VTransformComponent>>(this);```

VQuery Definition ```
template< class, class, class >
class VQuery;

template< class TOwned, class TGet = entt::get_t<>, class TExclude = entt::exclude_t<>>
class VQuery : public VQueryBase
{
public:
    // ...

    VQuery(VSystem* owningSys): VQueryBase(owningSys)
    {

        
    }

    
    void* operator new(std::size_t) = delete;
    void operator delete(void*) = delete;

    void InitializeQuery(VEntityManager& entity_manager) override
    {
        
        m_Group = entity_manager.group<TOwned, TGet, TExclude>();
    }
    
    auto& GetGroup()
    {
        entt::registry f;
        
        return m_Group;
    }

private:

    entt::basic_group<TOwned, TGet, TExclude> m_Group;
};```

Compile Errors : ```7>entt.hpp(18783,58): Error C2039 : 'base_type': is not a member of 'VTransformComponent'
7>entt.hpp(18783,58): Error C2146 : syntax error: missing '>' before identifier 'base_type'```

The line of code in entt.hpp
```using base_type = std::common_type_t<typename Owned::base_type..., typename Get::base_type..., typename Exclude::base_type...>;```

I have no clue about it, But it's on the right track I believe.

**Comment by Innokentiy-Alaytsev** [comment]
How do you instantiate your `VQuery` class?

**Comment by Diarica** [comment]
> VQuery<entt::owned_t<VTransformComponent>> m_TransformQuery = VQuery<entt::owned_t<VTransformComponent>>(this);

`VQuery<entt::owned_t<VTransformComponent>> m_TransformQuery = VQuery<entt::owned_t<VTransformComponent>>(this);`

Constructor`



    VQuery(VSystem* owningSys): VQueryBase(owningSys)
    {

        
    }`

**Comment by Innokentiy-Alaytsev** [comment]
Turns out it's [a bit more complicated](https://godbolt.org/z/6Pz9KfE5h) than this.

**Comment by Diarica** [comment]
> Turns out it's [a bit more complicated](https://godbolt.org/z/6Pz9KfE5h) than this.

100% works! You're my savior! Thank you!

**Comment by Diarica** [comment]
For people would likes to create wrappers, here is a example of entt::group, and you could create a empty group first and initialize it later. Thanks to @Innokentiy-Alaytsev !
Prerequisites: Base group class for type-erasure
```
class VQueryBase
{
public:
    void* operator new(std::size_t) = delete;
    void operator delete(void*) = delete;

    
    VQueryBase(VSystem* owningSys)
    {
        owningSys->RegisterQueryToSystem(this);
    }
    virtual ~VQueryBase()
    {
        m_OwningSys->UnRegisterQueryFromSystem(this);
    }
    

protected:
    VSystem* m_OwningSys;
private:
    virtual void InitializeQuery(VEntityManager& entity_manager)
    {
        
    }
    
    friend VSystem;
};
```
Type Alias of entt templates
```
template<typename... Type>
using Exclude_T = entt::exclude_t<Type...>;

template<typename... Type>
using Get_T = entt::get_t<Type...>;

template<typename... Type>
using Owned_T = entt::owned_t<Type...>;

```

Actual entt::group wrapper, this contains a bit complicated C++ templates
```
template< class... TOwneds, class... TGets, class... TExcludes >
constexpr auto get_group_type (
    entt::owned_t< TOwneds... >, entt::get_t< TGets... >, entt::exclude_t< TExcludes... >) ->
        entt::basic_group< 
            entt::owned_t< entt::registry::storage_for_type< TOwneds >... >,
            entt::get_t< entt::registry::storage_for_type< TGets >... >,
            entt::exclude_t< entt::registry::storage_for_type< TExcludes >... > >;

template< class TOwned, class TGet, class TExclude >
using group_type_t = decltype (get_group_type (
    std::declval< TOwned > (), std::declval< TGet > (), std::declval< TExclude > ()));


template< class, class, class >
class VQuery;

template< class TOwned, class TGet = Get_T<>, class TExclude = Exclude_T<>>
class VQuery : public VQueryBase
{
 public:
    VQuery(VSystem* owningSys): VQueryBase(owningSys)
    {
        
    }

    void* operator new(std::size_t) = delete;
    // Non-deleted operator delete() is required for the virtual destructor. 
    // Without virtual destructor Clang gives me problems with vtable.
    // void operator delete(void*) = delete;


    void InitializeQuery(VEntityManager& entity_manager) override
    {    
        m_Group = MakeGroup(entity_manager.GetNativeRegistry(),TOwned{});
    }

    auto& Group()
    {
        return m_Group;
    }

 private:
    group_type_t<TOwned, TGet, TExclude> m_Group;


    
    template< class... TOwneds >
    static auto MakeGroup(
        entt::registry& entity_manager,
        Owned_T< TOwneds... > i_owned) 
    {
        return entity_manager.group< TOwneds... > (TGet {}, TExclude {});
    }
};

```

**Comment by Diarica** [comment]
Example
```
class VTransformSystem : public VSystem
{
public:

    void OnStartup(VEntityManager& manager) override;
    void OnExecute(VEntityManager& manager, const VSystemExecutionContext& context) override;
    void RecalculateTransformMatrixIfDirty(VTransformComponent& transform);

    void SetPosition(VTransformComponent& tran, const glm::vec3& position);
    void SetRotation(VTransformComponent& tran, const glm::quat& rotation);
    void SetScale(VTransformComponent& tran, const glm::vec3& scale);
    
    static glm::mat4 CalculateTransformMatrix(const glm::vec3& position, const glm::quat& rotation, const glm::vec3& scale);

    VQuery<Owned_T<VTransformComponent>> m_TransformQuery = VQuery<Owned_T<VTransformComponent>>(this);
};

```
